### PR TITLE
Create toposort.cxx

### PR DESCRIPTION
**Pull request by AmineGherib**
A C++ implementation of a basic graph with a toposort function that uses DFS and a BFS method just for kicks.

### CODE REVIEW

File: graph/toposort.cxx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+#include <set>
+#include <map>
+#include <queue>
+#include <vector>
+
+template <typename T>
+class Graph {
+	std::map<T, std::set<T>> g;
+public:
+	void addEdge(T a, T b) {
```
Comment (Quinny) [inline graph/toposort.cxx:11]: These and all other parameters should be const T&.

File: graph/toposort.cxx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+	std::vector<T> toposort() {
+		std::set<T> visited;
+		std::vector<T> result;
+
+		for(auto const& v : g) {
+			if(visited.find(v.first) == visited.end())
+				dfs(visited, v.first, result);
+		}
+		return result;
```
Comment (Quinny) [inline graph/toposort.cxx:26]: Wont this be backwards?

File: graph/toposort.cxx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	void dfs(std::set<T>& visited, const T& v, std::vector<T>& result) {
```
Comment (Quinny) [inline graph/toposort.cxx:28]: This is specifically a post-order dfs and should be named accordingly.  It should also probably be private, it has no use to anyone else.

File: graph/toposort.cxx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		visited.insert(v);
+		for(auto const& e : g[v]) {
+			if(visited.find(e) == visited.end()) {
+				dfs(visited, e, result);
+			}
+ 		}
+		result.push_back(v);
+	}
+
+	void bfs(graph<T> g, T start) {
```
Comment (Quinny) [inline graph/toposort.cxx:38]: Its a little weird that this is a member function of the graph class, but also takes a graph as the argument?  Also, what is the use case of this function?

Comment (Quinny) [reply graph/toposort.cxx:38]: Does this even compile? "Graph" is a class, not "graph"

**Review by Quinny — CHANGES_REQUESTED** [review]

File: graph/toposort.cxx (hunk 4 above)
Comment (DavieV) [reply graph/toposort.cxx:38]: This function doesn't actually do anything currently, maybe make it take in std::function that performs an operation on the elements?

**Pull request closed without merging**